What's New 96 (#1566)
MicrosoftDocs/edge-developer: microsoft-edge/devtools-guide-chromium/whats-new/2021/11/devtools.md

For details, see [Inline and live issue analysis](../../../../visual-studio-code/microsoft-edge-devtools-extension.md#inline-and-live-issue-analysis).

<!-- ====================================================================== -->
## Sources tool now notifies you when sourcemaps can't be loaded

<!-- Title: Get notified when DevTools cannot load your sourcemaps correctly -->
<!-- Subtitle: The Sources tool now provides several places in the UI when DevTools can't fetch or parse your sourcemaps. -->

In Microsoft Edge version 96, the **Sources** tool now provides several indications in the UI when DevTools cannot load your sourcemaps.  In the **Page** tab in the **Navigator** pane in the **Sources** tool, files that DevTools can't load sourcemaps for have a warning<!--todo: capture/show--> icon in place of the file<!--todo: capture/show--> icon.  

Selecting a file with the warning icon opens the file in the Sources tool with an infobar<!--todo: capture/show--> indicating that DevTools either couldn't fetch the sourcemap correctly from the server or couldn't parse the sourcemap correctly.

> From the infobar, you can learn more about the issue by selecting the **Open in Issues** button.  The **Issues** tool then opens in the **Drawer** and provides information about how to resolve the issue and get your sourcemap loaded correctly in DevTools.

Reviewer: Should the "Open in Issues" button be shown in the image below? I don't see it in the current image.

:::image type="content" source="../../media/2021/11/source-map-not-found.png" alt-text="The Sources tool now notifies you in the Issues tab when sourcemaps can't be loaded.":::

> To prevent the infobar about sourcemaps from cluttering the **Sources** tool, select the **Don't show again** button.  To prevent issues related to sourcemaps from cluttering the **Issues** tool, clear the **Include third-party issues** checkbox in the **Issues** tool.  To learn more about how DevTools fetches and parses sourcemaps, see [Map preprocessed code to source code](../../../javascript/source-maps.md).

Reviewer: Similar comment as above - I don't see the "Don't show again" button in the image above.

# What's New in DevTools (Microsoft Edge 96)

The following sections list the announcements from the Microsoft Edge DevTools team.  To try the latest features of DevTools and the Microsoft Edge DevTools extension for Visual Studio Code, read these announcements.  To stay up to date with the latest and greatest features in your developer tools, download the [Microsoft Edge preview channels](https://www.microsoftedgeinsider.com/download) and [follow the Microsoft Edge DevTools team on Twitter](https://twitter.com/EdgeDevTools).

If you are on Windows, Linux, or macOS, consider using the Canary preview channel of Microsoft Edge as your default development browser.  The preview channels give you access to the latest features of Microsoft Edge DevTools.

<!-- ====================================================================== -->
## New DevTools UI available (in preview)

<!-- 1 -->

<!-- Title: New DevTools UI available (in preview) -->
<!-- Subtitle: A more minimal, modern UI is coming to Microsoft Edge DevTools. Enable the "Focus Mode" experiment to preview new UI features such a more compact toolbar that keeps DevTools uncluttered and better adapts to small window sizes. -->

> The Microsoft Edge DevTools team is experimenting with a new DevTools UI: **Focus Mode**, which provides a vertical Activity Bar for all of your tools.  This Activity Bar gives a quick view, to have two tools open simultaneously, and provides several customization and organization options to personalize your tools.

Reviewer 2: It is vertical by default, but you can also turn it horizontal, so I guess removing it might help.

Reviewer 3: Left requested revisions below.

> The Microsoft Edge DevTools team is experimenting with a new DevTools UI: **Focus Mode**, which provides a vertical Activity Bar for all of your tools.  This Activity Bar gives a quick view, to have two tools open simultaneously, and provides several customization and organization options to personalize your tools.

Reviewer 3: ```suggestion
The Microsoft Edge DevTools team is experimenting with a new DevTools UI: **Focus Mode**. Focus Mode reduces distractions and clutter with a more modern and simplified layout. The new Activity Bar allows you to pin your favorite tools in a horizontal or vertical toolbar to use screen space more efficiently.
```

To give this new UI a try in Microsoft Edge version 96, select **Settings** (![the Settings gear icon in DevTools](../../../media/settings-gear-icon-light-mode.png)) > **Experiments** > **Focus Mode**.

Starting with Microsoft Edge version 96, the experiment's checkbox is labeled **Focus Mode** rather than **Focus Mode and DevTools Tooltips**.

> This is an experiment, so the UI may change and evolve in future versions of Microsoft Edge.  We're eager to hear your feedback about this new DevTools UI.  Send us your feedback by tweeting [@EdgeDevTools](https://twitter.com/edgedevtools).  Or, with the **Focus Mode** experiment turned on, at the bottom of the **Activity Bar**, select **Help** (![the Help icon in the Activity Bar in Focus Mode](../../../media/help-icon-of-focus-mode.png)) > **Feedback**, to show the **Send feedback** window.
> 

Reviewer 3: ```suggestion
This UI is still in development and may change in future versions of Microsoft Edge.  We're eager to hear your feedback about this new DevTools UI.  Send us your feedback by tweeting [@EdgeDevTools](https://twitter.com/edgedevtools).  Or, with the **Focus Mode** experiment turned on, at the bottom of the **Activity Bar**, select **Help** (![the Help icon in the Activity Bar in Focus Mode](../../../media/help-icon-of-focus-mode.png)) > **Feedback** to show the **Send feedback** window.

```

:::image type="content" source="../../media/2021/11/focus-mode.png" alt-text="Focus Mode, including the Activity Bar.":::

<!-- video: see work item -->

See also:

> *  [Group tools together in Focus Mode](../02/devtools.md#group-tools-together-in-focus-mode) in _What's New In DevTools (Microsoft Edge 90)_.

Author: I added this link.  Please compare the present section to the linked section.
@hxlnt @zoherghadyali

Reviewer 3: This link should be removed because this functionality (tool grouping) no longer exists.

*  [DevTools: Focus Mode UI](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/DevTools/FocusMode/explainer.md) in the `MSEdgeExplainers` repo.

<!-- ====================================================================== -->
## Console can once again appear in the top and bottom panels

<!-- 2 -->

<!-- Title: Fix: Console can be quickly toggled in top or bottom panel -->
<!-- Subtitle: By popular demand, you can now easily show or collapse Console in the bottom panel without having to move the tool. -->

Since [Microsoft Edge version 87](../../2020/10/devtools.md#move-tools-between-top-and-bottom-panels), you've been able to move any tool between the top and bottom panels, including the Console tool.  However, we heard from developers that moving the Console tool repeatedly was cumbersome.  As of Microsoft Edge version 96, the default experience for the Console tool has been restored, allowing you to quickly switch between a full-screen view of the Console (in the top panel) and a split-screen view (with the Console displayed in the bottom panel), without moving the Console tool.

> When a tool other than the **Console** is selected in the main toolbar, such as the **Elements** tool, the bottom panel (the **Drawer**) is restored and displays the **Console** panel:
> 

Reviewer 3: ```suggestion
**Console** can be opened in the bottom panel (the **Drawer**) while another tool is open:

```

<!-- 3 -->

<!-- Title: Dockable browser screencast, device emulation, and live issue reporting, and in Microsoft Edge DevTools for Visual Studio Code -->
<!-- Subtitle: Display your web project inside the editor, simulate different devices, and get notified about issues with your code while you develop it. --> 

### Display your web project inside the editor, and simulate different devices

The Microsoft Edge DevTools for Visual Studio Code extension now features a dockable screencast and device emulation:

:::image type="content" source="../../media/2021/11/edge-devtools-for-vscode-toggle-screencast.png" alt-text="Click the screencast icon to view the browser inside of Visual Studio Code.":::

You can see your web project in a dedicated tab inside of Visual Studio Code, and also simulate various devices:

:::image type="content" source="../../media/2021/11/edge-devtools-for-vscode-simulated-iphone-red-boxes.png" alt-text="The screencast showing the current document in an emulated iPhone 5 in the correct size and with an emulated touchscreen interface.":::

<!-- videos: see work item -->

> For details about the dockable screencast, see [Changing to headless mode](../../../../visual-studio-code/microsoft-edge-devtools-extension.md#changing-to-headless-mode).  For details about device emulation, see [Device emulation in the screencast](../../../../visual-studio-code/microsoft-edge-devtools-extension.md#device-emulation-in-the-screencast).

Reviewer 2: Headless mode is a different thing than the screencast, I opt for removing the first sentence.

## Sources tool now notifies you when sourcemaps can't be loaded

<!-- 4 -->

<!-- Title: Get notified when DevTools cannot load your sourcemaps correctly -->
<!-- Subtitle: The Sources tool now provides several places in the UI when DevTools can't fetch or parse your sourcemaps. -->

In Microsoft Edge version 96, the **Sources** tool now provides several indications in the UI when DevTools cannot load your sourcemaps.  In the **Page** tab in the **Navigator** pane in the **Sources** tool, files that DevTools can't load sourcemaps for have a warning<!--todo: capture/show--> icon in place of the file<!--todo: capture/show--> icon.  

Selecting a file with the warning icon opens the file in the Sources tool with an infobar<!--todo: capture/show--> indicating that DevTools either couldn't fetch the sourcemap correctly from the server or couldn't parse the sourcemap correctly.

From the infobar, you can learn more about the issue by selecting the **Open in Issues** button.  The **Issues** tool then opens in the **Drawer** and provides information about how to resolve the issue and get your sourcemap loaded correctly in DevTools.

> * To prevent the infobar about sourcemaps from cluttering the **Sources** tool, select the **Don't show again** button.

Reviewer 4: I'm not really sure these 2 lines make sense as bullets in a list. I propose making these 2 sentences in one paragraph and then moving it down below the image. It's not must-have information but it is nice-to-have since we're not sure if this will cause undue clutter for users.

To clarify what I'm proposing: I'd add "To learn more about how DevTools fetches and parses sourcemaps..." as the last sentence in the paragraph.

:::image type="content" source="../../media/2021/11/source-map-not-found.png" alt-text="The Sources tool now notifies you in the Issues tab when sourcemaps can't be loaded.":::

<!-- video: see work item -->

To learn more about how DevTools fetches and parses sourcemaps, see [Map preprocessed code to source code](../../../javascript/source-maps.md).

<!-- ====================================================================== -->
## Opening source files in Visual Studio Code now integrates better with the Sources tool

<!-- 5 -->

<!-- Title: Open source files directly in Visual Studio Code from DevTools -->
<!-- Subtitle: The "Open source files in Visual Studio Code" experiment now works more intuitively with the Sources tool. -->

In previous versions of Microsoft Edge, turning on the **Open source files in Visual Studio Code** experiment had unexpected behavior when you had already been using the **Sources** tool in DevTools.  Setting breakpoints would direct you to Visual Studio Code or to documentation to configure the experiment correctly.

> In Microsoft Edge version 96, the **Open source files in Visual Studio Code** experiment now integrates better with the Sources tool.

Reviewer 4: ```suggestion
In Microsoft Edge version 96, the **Open source files in Visual Studio Code** experiment now integrates better with the **Sources** tool.
```

> *  If the Sources tool is open and you then turn on the **Open source files in Visual Studio Code** experiment, setting breakpoints or opening files now opens the Sources tool, rather than directing you to Visual Studio Code or to documentation to configure the experiment correctly.

Reviewer 4: ```suggestion
*  If the **Sources&& tool is open and you then turn on the **Open source files in Visual Studio Code** experiment, setting breakpoints or opening files now opens the **Sources** tool, rather than directing you to Visual Studio Code or to documentation to configure the experiment correctly.
```

   :::image type="content" source="../../media/2021/11/sources-tool-versus-open-in-vs-code.png" alt-text="If the Sources tool is open, setting breakpoints or opening files opens the Sources tool, even if you then turn on the 'Open source files in Visual Studio Code' experiment.":::

> *  As with earlier versions of Microsoft Edge, if the Sources tool isn't open in DevTools and you then turn on the **Open source files in Visual Studio Code** experiment, setting breakpoints or opening files from tools other than the Sources tool opens the file in Visual Studio Code.

Reviewer 4: ```suggestion
*  As with earlier versions of Microsoft Edge, if the **Sources** tool isn't open in DevTools and you then turn on the **Open source files in Visual Studio Code** experiment, setting breakpoints or opening files from tools other than the **Sources tool** opens the file in Visual Studio Code.
```
